src/mcMd/species/LinearSG.cpp

```diff
@@ -0,0 +1,399 @@
+         nAngle_ = nBond_ - 1;
+         if (nAngle_ > 0) {
+            loadParameter<int>(ar,"angleType", angleType_);
+         }
+      } else {
+         nAngle_ = 0;
+      }
+      #endif
+      #ifdef SIMP_DIHEDRAL
+      hasDihedrals_ = 0;
+      loadParameter<int>(ar,"hasDihedrals", hasDihedrals_, false);
+      if (hasDihedrals_) {
+         if (nAtom_ > 3) {
+            nDihedral_ = nAtom_ - 3;
+         } else {
+            nDihedral_ = 0;
+         }
+         if (nDihedral_ > 0) {
+            loadParameter<int>(ar, "dihedralType", dihedralType_);
+         }
+      } else {
+         nDihedral_ = 0;
+      }
+      #endif
+      buildLinear();
+      loadParameter<Pair <int> >(ar, "typeIds", typeIds_);
+      
+      ar & beadTypeIds1_;
+      ar & beadTypeIds2_;
+
+      loadParameter<double>(ar, "weightRatio", weightRatio_);
+      allocateSpeciesMutator(capacity(), 2);
+
+      // Set statistical weights
+      double sum = weightRatio_ + 1.0;
+      mutator().setWeight(0, weightRatio_/sum);
+      mutator().setWeight(1, 1.0/sum);
+   }
+
+   /*
+   * Save internal state to an archive.
+   */
+   void LinearSG::save(Serializable::OArchive &ar)
+   {
+      ar << moleculeCapacity_;
+      ar << nAtom_;
+      ar << typeIds_;
+      ar << beadTypeIds1_;
+      ar << beadTypeIds2_;
+      ar << weightRatio_;
+      #ifdef SIMP_ANGLE
+      Parameter::saveOptional(ar, hasAngles_, hasAngles_);
+      if (hasAngles_ && nAngle_ > 0) {
+         ar << angleType_;
+      }
+      #endif
+      #ifdef SIMP_DIHEDRAL
+      Parameter::saveOptional(ar, hasDihedrals_, hasDihedrals_);
+      if (hasDihedrals_ && nDihedral_ > 0) {
+         ar << dihedralType_;
+      }
+      #endif
+   }
+
+   /*
+   * Return NullIndex for every atom.
+   * Set initial typeIds
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateAtomTypeId(int index) const
+   { return 0; }
+
+   /*
+   * Return 0 for every bond.
+   *
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateBondTypeId(int index) const
+   { return 0; }
+
+   #ifdef INTER_ANGLE
+   /*
+   * Return 0 for every angle.
+   *
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateAngleTypeId(int index) const
+   { return 0; }
+   #endif
+
+   #ifdef INTER_DIHEDRAL
+   /*
+   * Return 0 for every dihedral.
+   *
+   * Used by Linear::buildLinear().
+   */
+   int LinearSG::calculateDihedralTypeId(int index) const
+   { return 0; }
+   #endif
+
+   /*
+   * Change the type of a specific molecule.
+   */
+   void LinearSG::setMoleculeState(Molecule& molecule, int stateId)
+   {
+      int nAtom  = molecule.nAtom();
+      for (int i = 0; i < nAtom; ++i) {
+         if (stateId == 0) {
+            beadIdentities_ = beadTypeIds1_;
+         } else {
+            beadIdentities_ = beadTypeIds2_;
+         }
+         molecule.atom(i).setTypeId(beadIdentities_[i]);
+      }
+      setMoleculeStateId(molecule, stateId);
+   }
+
+   /*
```

Comment:
The current Species interface doesn't have any functions to generate random molecules. There is 
instead a separate Generator base class. Subclasses of generators are created by a GeneratorFactory,
that (for now) will instantiate a LinearGenerator for any subclass of Linear.  The only two Generator
subclasses that know exist are a PointGenerator and a LinearGenerator.

Reply:
Sorry, I now see that these are commented out. Please remove commented out function declarations (in header) and definitions (here)